File: Breadth_first_search_algorithm/Bài_toán_8.py
from collections import deque

def bfs_simple_graph(adj, start):
    visited = set()
    queue = deque([start])
    order = []

    while queue:
        node = queue.popleft()
        if node not in visited:
            visited.add(node)
            order.append(node)
            for neighbor in adj[node]:
                if neighbor not in visited:
                    queue.append(neighbor)
    return order

graph = {
    0: [1, 2],
    1: [0, 3],
    2: [0, 3],
    3: [1, 2, 4],
    4: [3]
}

print("BFS traversal:", bfs_simple_graph(graph, 0))

File: Breadth_first_search_algorithm/Bài_toán_8.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

void bfs_simple_graph(const vector<vector<int>>& adj, int start) {
    int n = adj.size();
    vector<bool> visited(n, false);
    queue<int> q;
    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int u = q.front(); q.pop();
        cout << u << " ";
        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}

int main() {
    vector<vector<int>> graph = {
        {1, 2},
        {0, 3},
        {0, 3},
        {1, 2, 4},
        {3}
    };

    cout << "BFS traversal: ";
    bfs_simple_graph(graph, 0);
    cout << endl;

    return 0;
}
